docs/spec.md

```diff
@@ -2323,6 +2389,13 @@ More formally, `result[result_index] = dot_product`, where:
       dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
       body=lambda x, y: add(x, y))`.
   * `rounded_dot_product = round_nearest_even(integer_dot_product * (scale(reshaped_lhs_slice) * scale(reshape_rhs_slice) / scale(result)))`.
+    If the `scale(reshaped_lhs_slice)`, `scale(reshape_rhs_slice)`, and
+    `scale(result)` are respectively represented as integer multiplier and
+    shift pairs, like,
+    `(M_reshaped_lhs_slice, n_reshaped_lhs_slice), (M_reshaped_rhs_slice, n_reshaped_rhs_slice), and (M_result, n_result)`,
+    then the computation can be implemented as
+    `rounded_dot_product = (integer_dot_product * M + (1<<(n-1))) >> n`, where
+    `M = ((M_reshaped_lhs_slice * M_reshaped_rhs_slice) / M_result)) and n = (n_reshaped_lhs_slice + n_reshaped_rhs_slice - n_result)`.
```

Comment:
I think there are some issues with doing this integer M divide directly as stated. It can overflow if for example M_reshaped_lhs_slice = M_reshaped_rhs_slice = (1<<31)-1 and M_result=(1<<30) then M=(1<<32)-4 which is too large for int32 range. It could also underflow and should ideally be rounded. Since the resulting scale is M * 2^(-n), overflow can be avoided if n>1 by reducing n (which reduces M). Similarly, underflow can be avoided by increasing n (if not at maximum n). So, I would suggest writing this as n = (n_reshaped_lhs_slice + n_reshaped_rhs_slice - n_result), M = scaling_divide(M_reshaped_lhs_slice*M_reshaped_rhs_slice, M_result, &n) where this function divides and adjusts for overflow and underflow by altering n.

Reply:
Thanks for the elaborate write up! That is super helpful.

One follow up on a detail: Given that a floating-point scale is always positive, and the way `M` is calculated in the PR, I assume that M, the multiplier, will be in the range `0 <= M < 2^(W-1)` with `W`-bit  storage. With that we can skip the underflow case, right? Let me know if I missing something.

Reply:
Hi @dominicsymes 
Thanks for the above reply.
 
I think I still have a few open questions related to how one the internals of  the `scaling_divide`.
 
For the overflow case, 
Let us take same example, M_reshaped_lhs_slice = M_reshaped_rhs_slice = (1<<31)-1 and M_result=(1<<30)

Inside scaling_divide, if we  do the integer math with (M_reshaped_rhs_slice * M_reshaped_rhs_slice / M_result) we will get M = 4294967292. As the resulting scale is M * 2^(-n), we need to increase n. How is the value of n chosen? Is it by solving some inequality like Mx2^-n <= 2^31-1  or maybe some fixed value of n?


For the underflow case,if M_reshaped_lhs_slice = M_reshaped_rhs_slice = (1<<14) and M_result=(1<<30), M = 0.  
With that, the following integer computation is zero irrespective of the value on n. Maybe I am missing the math involved here. I want to better understand how  reducing n helps here.

 `rounded_dot_product = (integer_dot_product * M + (1<<(n-1))) >> n`

Reply:
Hi @sdasgup3 
The scale should be of the form M*2^(-n) where n is chosen so M does not overflow and has good precision. For example, to generate an M in range  0 <= M < (1<<31) the following should adjust n=(n_reshaped_lhs_slice + n_reshaped_rhs_slice - n_result) by decreasing to prevent overflow, increasing to prevent underflow:
- T = M_reshaped_lhs_slice * M_reshaped_rhs_slice
- while  T >= (M_result<<31) and n >= minimum_n then T=T/2; n=n-1;
- while  T < (M_result<<30) and n < maximum_n then T=T*2; n=n+1;
- M = T/M_result
- if n < minimum_n then saturate to maximum scale of M=(1<<31)-1, n=minimum_n

Reply:
Let me propose to limit this PR with (1) Introduction of multipliers and shifts in the type signature of `QuantizedTensorType` and (2) Specifying a conversion from floating-point scales to integer multipliers and shifts.  

We will address the  the specification of individual ops (like `dot_general`, `add`) with integers multipliers and shifts  in a separate PR.

Let me know.

Reply:
Hi @sdasgup3,

Your proposal makes sense. The discussion above is more specific to the operation (in this case dot_product) and I agree would be more easily handled in a separate PR relating to that operation.